module/icp/algs/impl/impl.c

```diff
@@ -0,0 +1,356 @@
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright 2021 Jan Kasiak. All rights reserved.
+ */
+
+/*
+ * Based on aes_impl.c
+ */
+
+#include <sys/simd.h>
+#include <impl/impl.h>
+
+static const struct {
+	char *name;
+	uint32_t sel;
+} alg_impl_opts[] = {
+	{ "cycle", ALG_IMPL_CYCLE },
+	{ "fastest", ALG_IMPL_FASTEST },
+};
+
+#define	BENCHMARK_NS MSEC2NSEC(2)
+
+#if defined(_KERNEL)
+
+/*
+ * Print benchmark stat headers.
+ *
+ * @param[out] buf
+ * @param      size of buffer
+ */
+static int
+impl_kstat_headers(char *buf, size_t size)
+{
+	ssize_t off = 0;
+
+	off += snprintf(buf + off, size, "%-17s", "implementation");
+	off += snprintf(buf + off, size - off, "%-15s\n", "bytes/second");
+
+	return (0);
+}
+
+static int
+impl_kstat_data(char *buf, size_t size, void *data)
+{
+	const alg_impl_ops_bandwidth_t *bw = data;
+
+	ssize_t off = 0;
+
+	off += snprintf(buf + off, size - off, "%-17s", bw->ops->name);
+	off += snprintf(buf + off, size - off, "%-15llu\n",
+	    (u_longlong_t)bw->bandwidth);
+
+	return (0);
+}
+
+/*
+ * Get the next kstat line data.
+ *
+ * @param[in,out] ksp
+ * @param         n
+ *
+ * @retval NULL end of data
+ * @return alg_impl_ops_bandwidth_t* from conf->supported
+ */
+static void *
+impl_kstat_addr(kstat_t *ksp, loff_t n)
+{
+	alg_impl_conf_t *conf = ksp->ks_private;
+
+	if (n == 0) {
+		return (&conf->bandwidth_fastest);
+	}
+
+	if (n <= conf->supported_n) {
+		return (void *)(conf->bandwidth + (n - 1));
+	}
+
+	return (NULL);
+}
+#endif
+
+void
+alg_impl_init(alg_impl_conf_t *conf)
+{
+	const alg_impl_ops_t *curr_impl, *fast_impl = conf->generic;
+	size_t i, c;
+	uint32_t max_priority = 0;
+
+	/* Copy from available to supported */
+	for (i = 0, c = 0; i < conf->available_n; i++) {
+		curr_impl = conf->available[i];
+
+		if (curr_impl->is_supported()) {
+			conf->supported[c] = curr_impl;
+			/* Ops for kstat, even if there is no benchmark. */
+			conf->bandwidth[c].ops = curr_impl;
+			c++;
+
+			/* Keep track of fastest */
+			if (curr_impl->priority > max_priority) {
+				max_priority = curr_impl->priority;
+				fast_impl = curr_impl;
+			}
+		}
+	}
+	conf->supported_n = c;
+
+	/* Run benchmark. */
+	if (conf->benchmark && kfpu_allowed()) {
+		/* Reset fastest assumption. */
+		uint64_t max_bw = 0;
+		fast_impl = conf->generic;
+
+		/* Allocate memory for benchmarking. */
+		size_t buffer_size = conf->benchmark_buffer_size;
+		uint8_t *buffer = vmem_alloc(buffer_size, KM_SLEEP);
+
+		/* Warmup. */
+		for (i = 0; i < conf->benchmark_buffer_size; i++) {
+			buffer[i] = i % 255;
+		}
+
+		/* Loop over supported implementations. */
+		for (i = 0; i < conf->supported_n; i++) {
+			const alg_impl_ops_t *ops = conf->supported[i];
+			uint64_t run_bw;
+			uint64_t run_count = 0;
+			uint64_t start, run_time_ns;
+
+			kpreempt_disable();
```

Comment:
I think this (and calls to them in general) need a `kfpu_begin()`/`kfpu_end()` wrapper if you're calling out into SIMD/FPU instructions, or you can wind up with problems like @adamdmoss is having.

See, for example, how the GCM code calls it before calling accelerated instructions, or the fletcher or raidz parity code likewise.

Reply:
K - I added the wrapping in dfa1f4299cc61f5b652768e3b866eb05ca9efb3b

Should I remove the calls to `kpreempt_disable` - if I use `kfpu_begin` instead will that be ok?

I got the `kpreempt_disable` from `module/zcommon/zfs_fletcher.c` where they do benchmarking.

Reply:
If you check, the definition of kfpu_begin is basically "`kpreempt_disable();` `[save FPU state]`", so you should be good doing that.

Reply:
Ah - now I see it.

I removed the redundant preempt calls from the benchmark portion.

Thanks!